text/0000-vec-fallible-allocation.md

```diff
@@ -0,0 +1,320 @@
+    pub fn try_into_boxed_slice(self) -> Result<Box<[T], A>, TryReserveError>;
+    pub fn try_push(&mut self, value: T) -> Result<(), TryReserveError>;
+    pub fn try_resize(&mut self, new_len: usize, value: T) -> Result<(), TryReserveError>;
+    pub fn try_resize_with<F>(&mut self, new_len: usize, f: F) -> Result<(), TryReserveError> where F: FnMut() -> T;
+    pub fn try_shrink_to(&mut self, min_capacity: usize) -> Result<(), TryReserveError>;
+    pub fn try_shrink_to_fit(&mut self) -> Result<(), TryReserveError>;
+    pub fn try_split_off(&mut self, at: usize) -> Result<Self, TryReserveError> where A: Clone; // NOTE: still panics if given an invalid index
+    pub fn try_with_capacity_in(capacity: usize, alloc: A) -> Result<Self, TryReserveError>;
+}
+
+#[doc(hidden)]
+pub fn try_from_elem<T: Clone>(elem: T, n: usize) -> Result<Vec<T>, TryReserveError>;
+
+#[doc(hidden)]
+pub fn try_from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Result<Vec<T, A>, TryReserveError>;
+
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Bifurcating the API surface like this is undesirable: it adds yet another decision point for developers using a fundamental type, and requires the developer to
+stop and understand what could possibly fail in the `try_` version. Additionally, this bifurcation can't stop at just `Vec`: What about the rest of the
+collection types? What about the traits that `Vec` implements (e.g., `IntoIter`)? What about other types that implement those traits? What about functions that
+use those traits? This bifurcation becomes viral both within the standard library and to other crates as well.
+
+Implementing both functions using a generic implementation method also complicates the code within `Vec`, making it harder to understand and maintain. While
+there is some complex code within `RawVec`, the code in `Vec` itself tends to be fairly straightforward and within reason for an inexperienced developer to
+understand.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Rationale
+Adding methods with the `try_` prefix follows the existing pattern within the standard library (as used by `Box::try_new` and `Vec::try_reserve`) and doesn't
+require any changes to the language or the compiler. Although this design does bifurcate the functions, it doesn't bifurcate the type: `Vec` is still `Vec` and
+so works with any crate expecting a `Vec`. Users of `Vec` can choose if they are in a path where allocation failures are recoverable (e.g., handling an
+incoming request), or non-recoverable (e.g., during application startup) without having to change the type they are using.
+
+If these methods are not added to the standard library, then it is highly likely that the code bases that require fallible allocation will take one of these
+approaches themselves, forking the standard library if required.
```

Comment:
Longer rational section emphasizing the key benefits of the current approach.

```suggestion
For the following reasons, we believe this the best option available at this time, and worthwhile to do rather than wait for a better option.
These are the reason why:

 - **Conforms to existing pattern**: Adding methods with the `try_` prefix follows the existing pattern within the standard library (as used by `Box::try_new` and `Vec::try_reserve`).
   They were added on a case-by-case basis for similar reasons. We are simply trying to push that effort further long without causing undue controversy.
 
 - **Works in Rust today**: It doesn't require any changes to the language or the compiler. Even if we knew exactly what change we would want (which we don't) coordinating the change and library modification afterwords is a long process.
   Until marquee use-cases like Rust in Linux are accepted upstream, we are hesitant to add uncertainty with high-latency projects.
  
 - **No new types splitting the ecosystem**: Although this design does bifurcate the functions, it doesn't bifurcate the type.
   `Vec` is still `Vec` and so works with any crate expecting a `Vec`.
   Multiple `Vec` types would hinder library interop when code one didn't write requires the other `Vec`.
   More functions / ad-hoc `try_` namespacing might be aesthetically displeasing, but doesn't levy any of these harsh forking penalties.

   Users of `Vec` can choose if they are in a path where allocation failures are recoverable (e.g., handling an incoming request), or non-recoverable (e.g., during application startup) without having to change the type they are using.

- **Usual beloved `Result` ergonomics**: Methods that requiring catching allocations failures more coarsely or out-of-band are more analogous to exception handling and errno, both of which are widely disliked by Rustaceans in comparison to `Result` and `?`.

- **Avoids forks, encourages interop** Conversely, if these methods are not added to the standard library, then it is highly likely that the code bases that require fallible allocation will take one of these approaches themselves, forking the standard library if required.
 Coalescing around a single fork across projects is unlikely, as it fostering a crates.io library ecosystem around such a fork.
 The longer we wait, the more low-level projects of this sort we are aiming for are likely to give up on reconcilling with the standard library, or sharing code with similar projects via libraries.
  
- **Roughly forwards compatible**: The ideal solution would avoid the `try_` ad-hoc namespacing without sacrificing these other benefits.
  While we don't know what that looks like on the *language* side, we are reasonably confident the *library* side will be quite similar.
  
  The essential insight of the current implementation is that we can share code between the old infallible and new fallible styles by using a polymorphic error type `Result<_, E>`.
  By making `E = !`, we "push" up the panicking as close to the original failure as possible, just like today, avoiding `Result`-propogating overhead when we're just going to end up panicking anyways.
  While we don't know the ideal form of this looks like *statically* in terms of types, coercions, and other compile-time features, we are pretty confident we do have what we want *dynamically* with the *behavior* of the old and new (`try_`) functions being what everyone wants.
  
  The current implementation provides both behaviors with minimal duplication, and so we suspect that whatever final form the interface takes, as long as it continues supporting these two desirable behaviors, our unified implementation an continue providing them.
```

Reply:
Thanks: I've merged this in with a few modifications.